ddtrace/contrib/vertica/patch.py

```diff
@@ -0,0 +1,265 @@
+    span.set_metric(dbx.ROWCOUNT, instance.rowcount)
+
+
+def cursor_after(instance, cursor, span, conf, *args, **kwargs):
+    tags = {}
+    tags[net.TARGET_HOST] = instance.options["host"]
+    tags[net.TARGET_PORT] = instance.options["port"]
+    if "user" in instance.options:
+        tags[dbx.USER] = instance.options["user"]
+    if "database" in instance.options:
+        tags[dbx.NAME] = instance.options["database"]
+
+    pin = Pin(
+        service=config.vertica["service_name"],
+        app=APP,
+        app_type=AppTypes.db,
+        tags=tags,
+        _config=config.vertica["patch"]["vertica_python.vertica.cursor.Cursor"],
+    )
+    pin.onto(cursor)
+
+
+# tracing configuration
+config._add(
+    "vertica",
+    {
+        "service_name": "vertica",
+        "app": "vertica",
+        "app_type": "db",
+        "patch": {
+            "vertica_python.vertica.connection.Connection": {
+                "routines": {
+                    "cursor": {"trace_enabled": False, "span_end": cursor_after}
+                }
+            },
+            "vertica_python.vertica.cursor.Cursor": {
+                "routines": {
+                    "execute": {
+                        "operation_name": "vertica.query",
+                        "span_type": sql.TYPE,
+                        "span_start": execute_span_start,
+                        "span_end": execute_span_end,
+                    },
+                    "copy": {
+                        "operation_name": "vertica.copy",
+                        "span_type": sql.TYPE,
+                        "span_start": copy_span_start,
+                    },
+                    "fetchone": {
+                        "operation_name": "vertica.fetchone",
+                        "span_type": "vertica",
+                        "span_end": fetch_span_end,
+                    },
+                    "fetchall": {
+                        "operation_name": "vertica.fetchall",
+                        "span_type": "vertica",
+                        "span_end": fetch_span_end,
+                    },
+                    "nextset": {
+                        "operation_name": "vertica.nextset",
+                        "span_type": "vertica",
+                        "span_end": fetch_span_end,
+                    },
+                }
+            },
+        },
+    },
+)
+
+
+def patch():
+    global _PATCHED
+    if _PATCHED:
+        return
+
+    _install(config.vertica)
+    _PATCHED = True
+
+
+def unpatch():
+    global _PATCHED
+    if _PATCHED:
+        _uninstall(config.vertica)
+        _PATCHED = False
+
+
+def _uninstall(config):
+    for patch_class_path in config["patch"]:
+        patch_mod, _, patch_class = patch_class_path.rpartition(".")
+
+        for patch_routine in config["patch"][patch_class_path]["routines"]:
+            mod = importlib.import_module(patch_mod)
+            cls = getattr(mod, patch_class)
```

Comment:
Same

Reply:
Also, I wonder if we should do:

```python
cls = getattr(mod, patch_class, None)
if not cls:
    continue
```

Just in case we end up with someone using a version of the library with a major API change.

Reply:
`unwrap` deals with this already.

Reply:
This is the case I am talking about.

```python
>>> m = type(sys)('test-module')
>>> getattr(m, 'unknown-class')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'test-module' has no attribute 'unknown-class'
```


Shouldn't be an issue patching via `wrapt`, but here we have a chance to access an attribute which doesn't exist.

Reply:
Oh, I see. Hmm, generally our stance has been that if the API changes then we no longer support the library/framework (we support the tested versions). But we could be a bit more graceful here, I'll change it up and add some logging messages.

Reply:
Yeah, I think the distinction I want to make is that because we are using `wrapt` to patch on import we likely won't fail on patching if the API/version doesn't match what we support.

However, this will very easily fail on `unpatch` if the API/version doesn't match what we support.